File: src/main/kotlin/com/leetcode/P3.kt
package com.leetcode

// https://github.com/antop-dev/algorithm/issues/110
class P3 {
    fun lengthOfLongestSubstring(s: String): Int {
        var ans = 0
        val sb = StringBuilder()
        for (ch in s) {
            if (ch in sb) {
                sb.delete(0, sb.indexOf(ch) + 1)
            }
            sb.append(ch)
            ans = maxOf(ans, sb.length)
        }
        return ans
    }
}

File: src/python3/com/leetcode/p3.py
# https://github.com/antop-dev/algorithm/issues/110
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        ans = 0
        _dict = {}
        anchor = 0
        for i, ch in enumerate(s):
            if s[i] in _dict:
                anchor = max(anchor, _dict[ch] + 1)
            ans = max(ans, i - anchor + 1)
            _dict[ch] = i
        return ans
